### PR TITLE
Add reading and writing of flux to full CBF files

### PR DESCRIPTION
**Pull request by phyy-nx**
Knock on effect from #488 is that reading and writing of flux to full CBFs changed nanoBragg simulations when they are instantiated using Beam models.  This caused a failure in simtbx/nanoBragg/tst_nanoBragg_cbf_write.py, which [sets](https://github.com/cctbx/cctbx_project/blob/master/simtbx/nanoBragg/tst_nanoBragg_cbf_write.py#L56) the flux using a beam object, write the simulated images, reads it back, re-simulates it, and compares the results.  The fix is to both read and write beam_flux in ImageCIF, as added in 2018 ImageCIF ([see v 1.7.11](http://www.bernstein-plus-sons.com/software/CBF/doc/cif_img_1.8.4.html)).

Will require a corresponding PR in cctbx_project.

### CODE REVIEW

**Comment by phyy-nx** [comment]
Corresponding PR: cctbx/cctbx_project#741.  The two PRs should be merged simultaneously.

File: src/dxtbx/format/cbf_writer.py [hunk 1]
```diff
@@ -209,6 +209,7 @@ def recursive_setup_basis_dict(key, parent_name="", panel_id=0):
                 diffrn_id,
                 False,
                 gain=[panel.get_gain() for panel in detector],
+                flux=beam.get_flux(),
```
Comment (graeme-winter) [inline src/dxtbx/format/cbf_writer.py:212]: Worth explicitly checking if `beam.get_flux()` is not None?

File: src/dxtbx/model/beam.py [hunk 2]
```diff
@@ -310,9 +310,20 @@ def imgCIF_H(cbf_handle):
             0.0,
         )
 
+        # and the flux if available
+        try:
+            cbf_handle.find_category(b"diffrn_radiation")
+            cbf_handle.find_column(b"beam_flux")
+            flux = cbf_handle.get_value()
+        except Exception as e:
+            if str(e).split()[-1] != "CBF_NOTFOUND":
```
Comment (graeme-winter) [inline src/dxtbx/model/beam.py:319]: Suspect that there are idiomatically better ways to catch this?

**Review by graeme-winter — APPROVED** [review]
Would be nice to catch the exceptions more nicely, and yeh I guess it is following the local style so unlikely to be improved.

File: src/dxtbx/format/cbf_writer.py (hunk 1 above)
Comment (phyy-nx) [reply src/dxtbx/format/cbf_writer.py]: That's checked in the function being called.

File: src/dxtbx/model/beam.py (hunk 2 above)
Comment (phyy-nx) [reply src/dxtbx/model/beam.py:319]: Maybe, but we are limited by the SWIG implementation here.  I looked for a 'check for column' function and didn't find it.  The find category/find column patterns seems to be what's available.  When these functions fail, in C they return an error code, but in python they raise a generic exception with a known string.

**Pull request merged**